Add HubSpot CMS features + Misc. changes

Sakib25800 opened the pull request:
### Description

This PR adds new HubSpot CMS features. This includes the ability to sync with the HubSpot Blog CMS and HubDB.

### Scopes
- Two new scopes need to be added: `hubdb`, `content`
- `hubdb` will need to be an configured as an optional scope meaning on the backend you should add a new query param `optional_scope` 

### Testing
- Since HubDB is a paid feature, you should use a HubSpot developer account
- [ ] HubDB Blog sync
  - [ ] Create a HubDB > Add some content > PUBLISH 
  - [ ] Launch the plugin via the CMS > Blog > Test usual Framer CMS plugin sync stuff 
- [ ] HubDB CMS sync
  - [ ] Launch the plugin via the CMS > HubDB > Select a table > Test usual Framer CMS plugin sync stuff
- [ ] Unauthorized states
  - [ ] Launch the plugin via the CMS logged out > Log in 
  - [ ] You should be redirected to the CMS menu page
- [ ] Clicking "Sync" on a unconfigured CMS should show the CMS menu
- [ ] Clicking "Sync" on a configured CMS should sync 
  
### Misc. Changes
- [ ] Nicer page routing animation 
- [ ] Scroll on meetings and forms page
- [ ] Nicer error page 

### Note
- We use the `name` property of the HubSpot CMS fields/columns as the ID of a collection field. 
- I would have given a cURL request to mock some data, but for some reason everything is resolving to error 500
- Remember to publish your HubDB table

alecmev commented on plugins/hubspot/src/hooks/useSearchParams.ts:
```diff
@@ -0,0 +1,37 @@
+        return Object.fromEntries(searchParams.entries()) as T
+    }, [searchString])
+
+    const setSearchParams = useCallback(
+        (newParams: Partial<T>) => {
+            const currentParams = new URLSearchParams(searchString)
+
+            // Update or add new params
+            Object.entries(newParams).forEach(([key, value]) => {
+                if (value === null || value === undefined) {
+                    currentParams.delete(key)
+                } else {
+                    currentParams.set(key, value)
+                }
+            })
+
+            const pathname = location.pathname
+            const newSearch = currentParams.toString()
+            const newUrl = `${pathname}${newSearch ? `?${newSearch}` : ""}`
+
+            setLocation(newUrl)
+        },
+        [searchString, setLocation]
+    )
```
This part can be dropped, looking at the usage.

alecmev commented on plugins/hubspot/src/hooks/useSearchParams.ts:
```diff
@@ -0,0 +1,37 @@
+
+    // Parse current search params
+    const params = useMemo(() => {
+        const searchParams = new URLSearchParams(searchString)
+        return Object.fromEntries(searchParams.entries()) as T
```
Maybe just straight up return `searchParams`? And use `.get` later? Less processing, and safer (types are lying here a bit).

Sakib25800 replied:
Yeah, that would be better.

alecmev reviewed (commented):
Great start! Some work is needed, but 90% of the way, the important basics work 👍 

1. I created a HubDB-based collection. Then in HubSpot, I added a new column (just simple text). When I go to "Manage" in Framer and tick the checkbox next to that column and apply, it doesn't get added, no matter what. If I create a new collection from the same HubDB - the column appears just fine.

2. When working with a HubDB collection, whether importing one for the first time, or managing, field checkboxes are rarely respected in full. Sometimes I uncheck 3 or 4 of them, and only 1 of them actually gets excluded. Can't quite pinpoint the pattern, but it's consistently misbehaving.

3. When it comes to HubSpot's API, I understand that [their client](https://github.com/HubSpot/hubspot-api-nodejs) is Node.js-only, but is there any reason not to use it at least for types? Maintaining them here is burden that's potentially avoidable.

4. Date+time in HubDB is ignored, but Framer supports it.

5. Doesn't seem like the video field gets populated.

6. I feel like there's some room for reuse in the field-mapping UI between blogs and HubDB, but I can see how it can get hairy. But, still, worth looking into. Abstracting them into something generic might pave the way toward more reuse across plugins, perhaps.

Sakib25800 commented on plugins/hubspot/src/blog.ts:
```diff
@@ -0,0 +1,169 @@
+
+    for (const [propertyName, propertyValue] of Object.entries(post)) {
+        if (propertyName === "slug") {
+            if (typeof propertyValue !== "string") continue
```
It can't happen, so I'll assert.

Sakib25800 commented on plugins/hubspot/src/App.tsx:
```diff
@@ -1,43 +1,217 @@
+            const isInCMSModes = mode === "syncManagedCollection" || mode === "configureManagedCollection"
+
+            if (isInCMSModes) {
+                const blogContext = await getBlogPluginContext()
+                const hubContext = await getHubDBPluginContext()
```
You're right - it should just be put inside.

Sakib25800 commented on plugins/hubspot/src/App.tsx:
```diff
@@ -1,43 +1,217 @@
+
+            if (blogContext.type === "update") {
+                navigate("/cms/blog")
+                setSize({ width: 340, height: 441 })
```
The `setSize` is used to set the initial size of the plugin. After the first page is determined the pages resize dynamically.

If I don't use `setSize` to set the initial width/height then the width/height go from `undefined` to the sizing of the page that has been loaded in.

What looks weird is when the sizing goes from `undefined` to a defined width/height - it looks like the plugin does some gymnastics.

Sakib25800 replied:
Yeah, that's probably better.

Sakib25800 commented on plugins/hubspot/src/api.ts:
```diff
@@ -80,6 +223,22 @@ interface RequestOptions {
+const PROXY_URL = "https://framer-cors-proxy.framer-team.workers.dev/?"
+const API_URL = "https://api.hubapi.com"
+
+export const queryKeys = {
+    blogPosts: (limit: number, properties: string[]) => ["blog-posts", limit, properties] as const,
+    publishedTables: (limit: number) => ["hubdb-tables", limit] as const,
+    publishedTable: (tableId: string) => ["hubdb-table", tableId] as const,
+    tableRows: (tableId: string, properties: string[], limit: number) =>
+        ["hubdb-table-rows", tableId, properties, limit] as const,
+    user: () => ["user"] as const,
+    account: () => ["account"] as const,
+    inboxes: () => ["inboxes"] as const,
+    forms: () => ["forms"] as const,
+    meetings: () => ["meetings"] as const,
+} as const
```
I've made them inline and turned off caching for the queries using `cachedFetch` using `staleTime: 0`.  

But From what I understand, `useQuery` and `cachedFetch` both have the same cache since they use the same `queryClient`?

Sakib25800 commented on plugins/hubspot/src/blog.ts:
```diff
@@ -0,0 +1,169 @@
+import { FieldsById, logSyncResult, MAX_CMS_ITEMS, SyncResult, SyncStatus } from "./cms"
+import { useMutation } from "@tanstack/react-query"
+
+export const PLUGIN_INCLUDED_FIELD_IDS = "hubspotPluginBlogIncludedFieldIds"
```
Flipped it. I guess in this case the exclusions will probably be smaller than inclusions. 

Though, also in this case, the (30) blog fields are fixed - which to store is a lot less than 2KB. So it probably won't  make much of a difference here even if a lot more blog fields were added.

jonastreub replied:
You can see which fields are not excluded because they have fields with matching ids. If you only want to know if anything changed you could also concatenate all hidden field ids and only store the hash of that value. Then it can never become more than the maximum size

alecmev replied:
Just in case, use this Google Sheets PR for inspiration: https://github.com/framer/plugins/pull/109

Sakib25800 commented on plugins/hubspot/src/App.tsx:
```diff
@@ -1,43 +1,217 @@
+            if (!isAuthenticated) {
+                handleNotAuthenticated()
+                return
+            }
```
Actually I take that back.

Sakib25800 commented on plugins/hubspot/src/App.tsx:
```diff
@@ -102,6 +102,8 @@ export function App() {
 
     const [, navigate] = useLocation()
 
+    const isAuthenticated = auth.isAuthenticated()
```
It's used in the `useEffect` hook so that if a user authenticates via the CMS we need to fetch all their plugin contexts.

alecmev replied:
Hmm. I guess I understand how that `useEffect` ends up being re-evaluated, because `useLocation` hook above likely fires when something authentication-related happens, and `isAuthenticated` gets updated too. But then probably best not do `auth.isAuthenticated()` again inside `getContexts`?

alecmev commented on plugins/hubspot/src/blog.ts:
```diff
@@ -32,12 +33,37 @@ export interface BlogPluginContextUpdate {
 
 export type BlogPluginContext = BlogPluginContextNew | BlogPluginContextUpdate
 
+/**
+ * Check if the current field configuration has changed.
+ */
+export function hasFieldConfigurationChanged(
+    currentManagedCollectionFields: ManagedCollectionField[],
+    allPossibleFieldIds: string[],
+    excludedFieldIds: string[]
+): boolean {
+    const currentIncludedFieldIds = currentManagedCollectionFields.map(field => field.id)
+    const expectedIncludedFieldIds = allPossibleFieldIds.filter(id => !excludedFieldIds.includes(id))
+
+    // Check if included field sets match
+    if (currentIncludedFieldIds.length !== expectedIncludedFieldIds.length) return true
+
+    // Compare the sets
+    const currentSet = new Set(currentIncludedFieldIds)
+    for (const id of expectedIncludedFieldIds) {
+        if (!currentSet.has(id)) return true
+    }
+
+    return false
+}
+
```
This is never used, no?

alecmev commented on plugins/hubspot/src/blog.ts:
```diff
@@ -46,10 +72,15 @@ export async function getBlogPluginContext(): Promise<BlogPluginContext> {
-    if (!excludedFieldIds) {
+    const includedFieldIds = collectionFields.map(field => field.id)
+    const excludedFieldIds = allPossibleFieldIds.filter(id => !includedFieldIds.includes(id))
+    const excludedFieldHash = simpleHash(excludedFieldIds.join(","))
```
Please, join like in Google Sheets, with something reasonably unique. It's unlikely to cause problems in blogs, but it's better to do it the same way everywhere. Should probably be a tiny utility function, used by both blogs and HubDB.

Probably also best to sort the ids, as to not rely on HubSpot's ordering.

alecmev reviewed (commented):
I think I'm missing something in HubDB's column hashing. The hashes are never compared? I was expecting 95% the same implementation as blogs, also hashing a list of excluded columns, and triggering "Manage" if the hashes don't match.

By the way, why the switch to column names instead of IDs?

Sakib25800 commented:
> Great start! Some work is needed, but 90% of the way, the important basics work 👍
> 
> 1. I created a HubDB-based collection. Then in HubSpot, I added a new column (just simple text). When I go to "Manage" in Framer and tick the checkbox next to that column and apply, it doesn't get added, no matter what. If I create a new collection from the same HubDB - the column appears just fine.
> 2. When working with a HubDB collection, whether importing one for the first time, or managing, field checkboxes are rarely respected in full. Sometimes I uncheck 3 or 4 of them, and only 1 of them actually gets excluded. Can't quite pinpoint the pattern, but it's consistently misbehaving.
> 3. When it comes to HubSpot's API, I understand that [their client](https://github.com/HubSpot/hubspot-api-nodejs) is Node.js-only, but is there any reason not to use it at least for types? Maintaining them here is burden that's potentially avoidable.
> 4. Date+time in HubDB is ignored, but Framer supports it.
> 5. Doesn't seem like the video field gets populated.
> 6. I feel like there's some room for reuse in the field-mapping UI between blogs and HubDB, but I can see how it can get hairy. But, still, worth looking into. Abstracting them into something generic might pave the way toward more reuse across plugins, perhaps.
> 7. Why refer to HubDB columns by name and not ID?

1. Should be solved now? 
2. Should be solved now? 
3. Yeah, I should have and I've now used as many as the types from their client as possible. Though there are slight differences between the [API guide](https://developers.hubspot.com/beta-docs/guides/api/cms/hubdb) and the types from the SDK - not really an issue though. 
4. Added - forgot that date **+ time** was added to the Framer CMS 
5. Added - we now need another scope: `files` since the API returns the `fieldId`. It can be required since it's not a paid feature. 
6. Abstracted into `FieldMapper.tsx`
7. The HubSpot API returns row data using column API names as keys (e.g., `{"col1ApiName": "value1"}`), not column IDs. Even though column definitions include both IDs and API names (e.g., `{"id": "123", "name": "col1ApiName"})`, row values only reference API names. Using column IDs would break field mapping, so we use API names to prevent data loss.

Something to note: a user can change the name of a column via HubSpot but there's a little indicator that tells them that the API uses this - not too sure if we can do much here? Also to note, if a user names their column something like "column" and then renames (the label) to "columz", HubSpot will keep the (API) name as "column".

alecmev commented:
Great progress, thanks! Did some testing:

1. Fixed.
2. Fixed.
3. Nice! Shame that their types are [so-so](https://github.com/HubSpot/hubspot-api-nodejs/blob/18c7c3816cde469463f75888c5d9d2fc985ff44c/codegen/cms/hubdb/models/HubDbTableRowV3.ts#L31).
4. The time part of "date and time" still doesn't work.
5. I updated the scopes, but the video field still doesn't get populated.
6. Nice!
7. Alright, as long as column IDs are stable. I do understand correctly that they're immutable on HubSpot side?
8. When importing a blog, all fields are unselected by default, is that intended?
9. In blogs it's possible to skip "slug", but it gets imported anyway. Mark it as "required" in the field mapper?

alecmev commented on plugins/hubspot/src/cms.ts:
```diff
@@ -0,0 +1,139 @@
+    currentFields: ManagedCollectionField[]
+    allPossibleFieldIds: string[]
+    storedHash: string
+    mode: "included" | "excluded"
```
Is the `excluded` mode used anywhere? I'd just drop it if not.

alecmev commented:
4. My bad, it actually isn't, filed an internal issue about it! I think it's fine as-is for now then, just keep the date part.
5. Hmm, there's no access to the video at all? Not a big deal for the first release, but would be nice to know for sure.
7. Can we not refer to columns by ID on Framer side, and translate IDs to current column names when we parse the column data provided by HubSpot?
8. Cool!
9. By didn't make sense to include you mean that it makes sense to always have it by default?
10. Thanks!
11. Re reference fields, sounds good, but if it's complex - better save it for a separate PR, we'd like to release soon.

alecmev commented:
4. Ah, can you undo the last commit, please? Let's let it work, just without the time component.
5. Alright! We can investigate deeper later, if somebody asks.
6. Okay, I was thinking that maybe we could silently rename if the ID remains the same, but it's probably better to be transparent and pop a field-mapping window no matter what, so might as well use the names 👍 

Great work! I think this is ready for merging, right after somebody else's QA.

Sakib25800 commented:
re: field mapping by id 

I now realise what you mean and silent renaming would definitely be better UX. I completely misunderstood 🤦‍♂️.

I'll get that done asap.

alecmev commented:
Oh, no worries re ID, I was more spitballing, but thank you for implementing this! Will test in a short bit.

alecmev commented:
Meant to ask, since videos didn't work out, is the `files` scope still needed?

Sakib25800 commented:
> Meant to ask, since videos didn't work out, is the `files` scope still needed?

Nope.

huntercaron commented:
I'm getting some weird results when syncing. For example, the "name" field never seems to sync properly, although the data is going into the slug field. For the rich text content, I can check the content field in the manage UI but it doesn't import, and doesn't remain checked. Any ideas?

<img width="2560" alt="Screenshot 2024-11-19 at 14 41 22" src="https://github.com/user-attachments/assets/f98d7381-b1d0-439f-ac1a-0ceeb6c6fef7">
![Uploading Screenshot 2024-11-19 at 14.41.39.png…]()

Sakib25800 commented:
> > @huntercaron, issue should be fixed.
> 
> I am still seeing the issue, only the checkbox persisting seems to be fixed. the name field and content field still aren't importing for me

Did you try again from the beginning? As in, from the CMS menu > create new HubDB collection > Testing. That Framer CMS collection you've shown in the image should be deleted - it won't work anymore, there was an error in the syncing logic.

huntercaron commented:
yup this is from the beginning, still seeing it

Side note, we've started auto-selecting the first item here, would be nice to add here as well.
<img width="418" alt="Screenshot 2024-11-19 at 15 53 23" src="https://github.com/user-attachments/assets/015f2bce-bf1c-4085-90d1-dfe8de88f724">

huntercaron commented:
I don't know hubspot deeply, but testing these features in a general sense it seems to work as expected!

alecmev commented:
Awesome! @Sakib25800, can you squash again (since you did it before), and rebase, please?
